### ISSUE TITLE
Search issue report 3

### ISSUE THREAD

**Issue by bo-io-gov-mo** [bo-io-gov-mo/website#28]
I open a new issue follow #26

mark down: for adjust the minor sort order of same diplom type, for example put "basic law" on top, I use **bo2004.Visits_cn** as **diplomlvl** provisional, only on A record.

later we need to do: 
1.change a suitable field name. 
2.update new record level no from diplomTypes table.

**Comment by xiquinho** [comment]
I can update new record level no from diplomTypes table easily but to change the field name I will need to recreate the replication on all servers, expect same errors.

**Comment by bo-io-gov-mo** [comment]
agree not change field name at this time, may be do it later.

**Comment by xiquinho** [comment]
done, new records will have diplomlvl  on field bo2004.Visits_cn

**Comment by bo-io-gov-mo** [comment]
http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E6%BE%B3%E9%96%80%E9%9B%BB%E5%8A%9B%E6%9C%89%E9%99%90%E5%85%AC%E5%8F%B8
![image](https://cloud.githubusercontent.com/assets/19220046/17467456/f1871176-5d50-11e6-8867-43c071246fba.png)

Please adment the search result line space and text color as below:
http://192.168.3.63/cn/search/?q=%E6%BE%B3%E9%96%80%E9%9B%BB%E5%8A%9B%E6%9C%89%E9%99%90%E5%85%AC%E5%8F%B8
![image](https://cloud.githubusercontent.com/assets/19220046/17467489/29679c1e-5d51-11e6-9a9b-6f3e58313f95.png)

**Comment by bo-io-gov-mo** [comment]
Search keyword auto-complete not implement yet.

**Comment by bo-io-gov-mo** [comment]
Modify chinese word:
http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E6%BE%B3%E9%96%80%E9%9B%BB%E5%8A%9B%E6%9C%89%E9%99%90%E5%85%AC%E5%8F%B8
![image](https://cloud.githubusercontent.com/assets/19220046/17468861/7d3f505e-5d5e-11e6-8bfa-f37f32e57ba7.png)

1.全文  2.搜尋
http://192.168.3.63/cn/search/?q=%E6%BE%B3%E9%96%80%E9%9B%BB%E5%8A%9B%E6%9C%89%E9%99%90%E5%85%AC%E5%8F%B8
![image](https://cloud.githubusercontent.com/assets/19220046/17468824/3809f660-5d5e-11e6-8dc1-56c892c26c39.png)

**Comment by bo-io-gov-mo** [comment]
Please display "show sql" provisionally for me confirm syntax and optimize sql db index.

**Comment by bo-io-gov-mo** [comment]
please show the result record count and page count on the top bar.
![image](https://cloud.githubusercontent.com/assets/19220046/17469582/17ac4a18-5d67-11e6-84bf-bcab25a89c58.png)


**找到9999項結果，瀏覽:1/99**

**Comment by xiquinho** [comment]
done, except for autocomplete, pl check.

**Comment by bo-io-gov-mo** [comment]
Please display "show sql" provisionally, it appear yesterday and disappear today.

**Comment by bo-io-gov-mo** [comment]
I want adjust many minor layout/margins/space/color things adjust on the search page, it may be adjust many times until finalize. I write down it to you modify, or I can adjust it by myself, please let me know how is convenience to you and cooperation together.

**Comment by xiquinho** [comment]
I suggest finish first all requirements, all tests and when is bug free adjust the minor layout/margins/space/color things.

If you want to adjust by yourself, you need to warn me first, I need to check-in the page, you will need to check-out and only after that make any changes. that's the only  safe way to make changes on the same page at the same time.

**Comment by xiquinho** [comment]
done, pl. check.

**Comment by xiquinho** [comment]
In some situations DatePublished is not enough to sort the records correctly, pl. add: 
DipOrder , 
DiplomNumber 
to your view to fix this.

**Comment by bo-io-gov-mo** [comment]
Already add DipOrder field to vwIndexQuery view. A, B and C table is ok, but DintDocList table can't find it, I put it zero.

 Others, When I take a look into DintDocList union query, I found below condition at the end: 
``
... AND Dint_DocList_DocLang_join.LangID = 22
``
I don't know is it correct?

````
...
UNION ALL
SELECT     DintDocList.tboID AS id, 'D' AS cat, DocName_pt AS DiplomNamept, DocDesc_pt AS DiplomDescpt, DocName_pt + DocDesc_pt AS Textpt, 
                      DocName_cn AS DiplomNamecn, DocDesc_cn AS DiplomDesccn, DocName_cn + DocDesc_cn AS Textcn, DintDocList.LinkID, DocDate AS DatePublished, 
                      '0' AS DiplomUrl, BONumber, BOYear, Serie, suplemento, 600 AS Diplomlvl, DintDocList.PID, DintDocList.PIDOriginal, 'Direito Internacional' AS CatNamePT, 
                      N'國際法' AS CatNameCN, 
                      0 AS typeID, 0 as DipOrder
FROM         DintDocList INNER JOIN
                      Dint_DocList_DocLang_join ON DintDocList.LinkID = Dint_DocList_DocLang_join.LinkID INNER JOIN
                      tbo2004 ON Dint_DocList_DocLang_join.LinkIDtbo2004 = tbo2004.LinkID
WHERE     tbo2004.Active = 1 AND MSAR = 1 AND Dint_DocList_DocLang_join.LangID = 22

````

**Comment by xiquinho** [comment]
DintDocList no need this is to sort only when the records have numbers, like Law 1/2016, 2/2016 and so on. DintDocList is OK to put all as zero.

AND Dint_DocList_DocLang_join.LangID = 22 , was for some testing only, pl. remove.

**Comment by bo-io-gov-mo** [comment]
but when i remove, it come out many duplicate record.

**Comment by xiquinho** [comment]
You need to remove it otherwise your view will only include records with PT version and now we have some records with CN and/or EN version only. I will deal with the record duplication on the code but to fix and test it I need you remove this filter from the view first. Otherwise, I never see any duplication.

**Comment by bo-io-gov-mo** [comment]
already remove that condition. pls take a look.

**Comment by bo-io-gov-mo** [comment]
now the view is exclude D, it only contain A+B+C, pls test it.

**Comment by bo-io-gov-mo** [comment]
I already created 4 index name **"_tfs"** in four tables: "**tbo2004**", "**tbo20002s**", "**tboanotariais**" and "**search**" table for optimize search query.  keep looking its performance.

**Comment by xiquinho** [comment]
DO NOT create any index or any other design changes on bo2004 databases (beta and production) without talk with me first, OK? All the index need to improve performance are already there is no need for more.

**Comment by xiquinho** [comment]
Beside what performance do you intend to improve with the "_tfs" index? I can't see any...

**Comment by bo-io-gov-mo** [comment]
I created it on bo2004_search, it is non-clustered index, it is some missing index report by estimates execution plan, these index is for first tab search query,  for example:

````
/*
Missing Index Details from SQLQuery4.sql - control.Bo2004_search (useri (250))
The Query Processor estimates that implementing the following index could improve the query cost by 80.448%.
*/

/*
USE [Bo2004_search]
GO
CREATE NONCLUSTERED INDEX [<Name of Missing Index, sysname,>]
ON [dbo].[tbo20002s] ([CatId],[active])
INCLUDE ([LinkID],[DiplomNamecn],[DiplomDesccn],[DipOrder],[BoNumber],[suplemento],[DatePublished],[BoYear],[Serie],[DiplomUrl])
GO
*/
````

If you really want to know improve or not, it is good to build a timer to record and compare the execution time. I don't know how to see.

**Comment by xiquinho** [comment]
I create two timers, one for the query itself, other for the filters. Regarding the "_tfs" index, what I don’t understand is how the performance will improve including these columns in the index, for example, BoNumber, as we never query the database by BoNumber. But Ok, if you are sure it will forget what I said.

**Comment by bo-io-gov-mo** [comment]
If you have those timer, can show it to result pages provisionally like "show sql" for me testing. 
Moreover, I found those tables missing a cluster-index, which I known is important for read/write performance. I'm not well known of this mechanism yet, but it is basic of db architecture. you can search information detail about this everywhere, like:
http://stackoverflow.com/questions/11787797/should-every-user-table-have-a-clustered-index
http://www.sql-server-performance.com/2007/clustered-indexes/

**Comment by xiquinho** [comment]
" I found those tables missing a cluster-index"

what tables?

**Comment by bo-io-gov-mo** [comment]
"tbo2004", "tbo20002s", "tboanotariais"

**Comment by xiquinho** [comment]
I don’t know well this topic also but I remember I already create all the indices needed for the old search and I never notice any performance problems. All those tables have an index on the primary key, I am not sure of difference it makes if is clustered or not.

Of course in the new search thinks look like a bit slow but I think is because the complexity of SQL query. 

Anyway, If you need to modify something for the new search just be sure what are you doing if you are going to modify something on the production databases.

**Comment by bo-io-gov-mo** [comment]
I saw the timer now, ty! I have some question understand, when I search 2 keyword case, the query timer similar but page display time big different, as below:

case 1: http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E9%81%8A%E8%89%87
Query took 640 Milliseconds - Filters took 343 Milliseconds
page load: 3 seconds ([use chrome pageload extension](https://chrome.google.com/webstore/detail/page-load-time/fploionmjgeclbkemipmkogoaohcdbig))

case 2: http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E6%BE%B3%E9%96%80
Query took 575 Milliseconds - Filters took 452 Milliseconds
page load: 18 seconds

It is in fact on testing, I don't know why, is the timer only count query time or page all execution time?where is the bottleneck?

**Comment by bo-io-gov-mo** [comment]
http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E9%81%8A%E8%89%87
![image](https://cloud.githubusercontent.com/assets/19220046/17581734/28139896-5fda-11e6-9566-70c684b9b394.png)

http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E6%BE%B3%E9%96%80
![image](https://cloud.githubusercontent.com/assets/19220046/17581711/f5150c86-5fd9-11e6-819b-9defaed9a3f2.png)

**Comment by xiquinho** [comment]
the timer only count the time between send the query to sql server and the results return from the sql server.

**Comment by xiquinho** [comment]
anyway, at this time nothing is optimized yet as I need to be sure about the final design before making any optimizations. I am planning to cache everything I can but right now I am not caching nothing as I need to see all the results without any cache. However the way this filters are designed it is hard to cache anything...

**Comment by bo-io-gov-mo** [comment]
so the bottleneck is not seems at sql server or sql query. pls check the web server execution process. The mainly difference of above is one return about 195 records and another return 91492 records, but I think it should be a number, needn't retrieve all records beside page display. This is only my suspect,  may be not correct.

**Comment by bo-io-gov-mo** [comment]
As I know cache can make performance better, it is true. But it makes harder to found the issue to solve, it is hide the problem.

**Comment by bo-io-gov-mo** [comment]
The text change of chinese not correct:
1. 全文 (now is correct)
2. 標題 
3. 搜尋

![image](https://cloud.githubusercontent.com/assets/19220046/17582432/38741e1e-5fde-11e6-9192-8cf584c8d1e7.png)

**Comment by xiquinho** [comment]
these are not optimized yet, I still have code duplications to delete and many things to optimize. But I need to know first what do you want for sure. When you tell me OK, this is final I will optimize everything, OK? Or not OK?

**Comment by bo-io-gov-mo** [comment]
Yes, It is ok **basically**. 
 beside diplomlvl not add to rank yet, which order is different:

TAB ALL
http://betan.io.gov.mo/cn/search/?sc=0&tab=&q=%E5%9F%BA%E6%9C%AC%E6%B3%95

TAB A
http://betan.io.gov.mo/cn/search/?sc=0&tab=a&q=%E5%9F%BA%E6%9C%AC%E6%B3%95&n=&ya=

**Comment by bo-io-gov-mo** [comment]
and if user click change the tab, can re-search with the same keyword in those tab data scope. or just reset it? now click tab is only change the tab options, but the result may be not belong the tab.

**Comment by xiquinho** [comment]
"beside diplomlvl not add to rank yet"

Where is missing?

**Comment by bo-io-gov-mo** [comment]
Regard to the **diplomlvl** of Tab A, I found the most simply modify sql query is:

http://betan.io.gov.mo/cn/search/?sc=0&tab=a&q=%E5%9F%BA%E6%9C%AC%E6%B3%95&n=&ya=
`````
With result_set As (Select ROW_NUMBER() OVER ( ORDER BY **title_Rank** DESC, text_Rank DESC, DatePublished DESC, DipOrder DESC ) As [row_number], FT_TBL.DiplomNamecn AS DipName, FT_TBL.DiplomDesccn AS DipDesc, FT_TBL.LinkID, FT_TBL.DatePublished, FT_TBL.DiplomUrl, FT_TBL.BONumber, FT_TBL.BOYear, FT_TBL.Serie, FT_TBL.pagina, FT_TBL.suplemento, 'a' as Tab, KEY_TBL.Title_RANK as RANK FROM tbo2004 AS FT_TBL INNER JOIN ( SELECT SUM(t.Title_Rank) AS Title_RANK, SUM(t.text_Rank) AS Text_RANK, s.tboID, s.LinkID, s.tab FROM ( SELECT [KEY], Rank as Title_Rank, 0 AS Text_Rank FROM CONTAINSTABLE(search, title_cn, N'"基本法"',language 1033) UNION ALL SELECT [KEY], 0 AS Title_Rank, Rank as Text_rank FROM CONTAINSTABLE(search, text_cn, N'"基本法"',language 1033) ) as t inner join Search as s on t.[key]=s.tboID GROUP BY s.tboID, s.LinkID, s.tab ) AS KEY_TBL ON (FT_TBL.LinkID=KEY_TBL.LinkID and KEY_TBL.tab='A') WHERE FT_TBL.Active = 'True' AND FT_TBL.DipDisplay = 1 ) Select * FROM result_set WHERE [row_number] BETWEEN 0 And 24
````

Try to replace the first **title_Rank**  text to:
**(Case When KEY_TBL.title_Rank > 0 Then KEY_TBL.title_Rank+FT_TBL.Visits_cn*100 Else KEY_TBL.title_Rank End)** 

this will cause sort order add diplomlvl with title_Rank. 
(we use Visits_cn to store diplomlvl)

**Comment by xiquinho** [comment]
"Regard to the diplomlvl of Tab A"

Only change in tab A? pl. confirm.

**Comment by bo-io-gov-mo** [comment]
yes, because you told me only A have diplomlvl.

**Comment by xiquinho** [comment]
Just to clarify in case you misunderstood, the intention of diplomlvl was to force some groups of records appear first when they have the same rank, for example DESPCE should came before DESPSAJ and DESPSAJ came before DESPSEF and so on. It was never my intention to manipulate the results on individual level just like you are doing now.

**Comment by bo-io-gov-mo** [comment]
Yes,  I know your intention of it.  I turn it to record level now for ordering, it is based on diplomlvl group level.  I need another field to store this record level. Now it is call Visits_cn on each table. Thanks for clarify for you and me know better.

**Comment by bo-io-gov-mo** [comment]
Another question: Visits_cn, Visits_pt, Visits_en is design for store access times of record link. It is very good for tune search result priority, why not use them ago or give up before?

**Comment by xiquinho** [comment]
“Another question”

Tbo2004 on production must be read only and it is not good idea give web apps permission to write on Tbo2004 beta. It is dangerous, resource intensive and cause a lot of stress on the server update the counters every time every thousands user visits thousands pages. Therefore, I had give up with the idea.

**Comment by xiquinho** [comment]
Btw, on tab B the default sort is RANK (as it should be) or DATE (as it is now) ?

**Comment by bo-io-gov-mo** [comment]
It is good question, it look not consist with others, reason may be they consider on user aspect or those data speciality. I will ask them again to confirm, and keep existing default until they request change it.

**Comment by xiquinho** [comment]
I found the sql query to count the number of records takes a large amount of time, not sure why it was written by you or Samuel. pl take a look and see if you can change something.

**Comment by xiquinho** [comment]
can you open a separate issue for each problem/question pl? It is very hard to find what is done and what is not like this. ty!

**Comment by bo-io-gov-mo** [comment]
I already separate these one by one.
